python/nammu/controller/AtfAreaController.py

```diff
@@ -122,6 +122,35 @@ def set_validation_errors(self, validation_errors):
         '''
         self.validation_errors = validation_errors
 
+    def undoOrRedo(self, forward):
+        """
+        Helper function for undo() and redo().  The boolean argument `forward`
+        specifies if the edit is to be undone (`False`) or redone (`True`).
+        """
+        currentEdit = None
+        try:
+            # Before actually redoing/undoing, get the edit about to be
+            # redone/undone, in order to then move focus to its pane.
+            if forward:
+                currentEdit = self.undo_manager.editToBeRedone()
+                self.undo_manager.redo()
+            else:
+                currentEdit = self.undo_manager.editToBeUndone()
+                self.undo_manager.undo()
+        except (CannotUndoException, CannotRedoException):
+            # These exceptions indicate we've reached the end of the edits
+            # vector.  Nothing to do
+            pass
+        else:
+            if currentEdit:
+                # Move focus to the pane where `currentEdit` was done.
+                editDoc = currentEdit.firstEdit().getDocument()
+                if editDoc == self.arabic_area.getStyledDocument():
+                    self.arabic_area.requestFocusInWindow()
+                elif editDoc == self.edit_area.getStyledDocument():
+                    self.edit_area.requestFocusInWindow()
```

Comment:
If we're using the secondary area and we undo, this will change the focus to the primary area. That's not ideal, but not sure how easy to fix (if the two areas share a styled document, I guess we can't distinguish that way).

If we end up having to look at more than the two areas we are checking at the moment, it may be better to check in a loop. This is probably better placed in a new PR, though - I'll open an issue about it.

Reply:
Opened issue #402 for this.